File: 프로그래머스/lv2/17680. ［1차］ 캐시/［1차］ 캐시.py
def solution(cacheSize, cities):
    answer = 0
    
    cache = {}
    
    idx = 0
    
    if cacheSize == 0:
        return len(cities) * 5
    
    
    for city in cities:
        city = city.lower()
        if city in cache:
            cache[city] = idx
            idx -= 1
            answer += 1
            
        else:
            if len(cache) >= cacheSize:
                maximum = -1000000
                maximumKey = ""
                for key in cache:
                    if maximum < cache[key]:
                        maximum = cache[key]
                        maximumKey = key
                
                del(cache[maximumKey])
                
            cache[city] = idx
            idx -= 1
            
            answer += 5
    
    return answer

File: 프로그래머스/lv2/17680. ［1차］ 캐시/［1차］ 캐시.java
import java.util.*;

class Solution {
    public int solution(int cacheSize, String[] cities) {
        int answer = 0;
        
        HashMap<String, Integer> cache = new HashMap<>();
        
        int idx = 0;
        
        if(cacheSize == 0) {
            return cities.length * 5;
        }
        
        for(String city : cities) {
            city = city.toLowerCase();
            
            if(cache.containsKey(city)) {
                cache.put(city, idx);
                idx--;
                answer++;
            } else {
                if(cache.size() >= cacheSize) {
                    int maximum = -1000000;
                    String maximumKey = "";
                    
                    for(String key : cache.keySet()) {
                        if(maximum < cache.get(key)) {
                            maximum = cache.get(key);
                            maximumKey = key;
                        }
                    }
                    
                    cache.remove(maximumKey);
                }
                
                cache.put(city, idx);
                idx--;
                
                answer += 5;
            }
        }
        
        return answer;
    }
}